### PR TITLE
Make InProcessExecutor use a plain Input instead of bytes

### PR DESCRIPTION
**Pull request by maxammann**
This PR alllws the usage of abstract non-binary inputs for the InProcessExecutor. This can be used for ASTs for example.

I hope I found all references, if not I think the CI should fail.

### CODE REVIEW

File: fuzzers/baby_fuzzer/src/main.rs [hunk 1]
```diff
@@ -26,7 +27,8 @@ fn signals_set(idx: usize) {
 #[allow(clippy::similar_names)]
 pub fn main() {
     // The closure that we want to fuzz
-    let mut harness = |buf: &[u8]| {
+    let mut harness = |input: &BytesInput| {
```
Comment (domenukk) [inline fuzzers/baby_fuzzer/src/main.rs:30]: Couldn't this be `I: HasBytesVec` then? Or does that get more complex

File: fuzzers/libfuzzer_libmozjpeg/src/lib.rs [hunk 2]
```diff
@@ -128,7 +129,8 @@ fn fuzz(corpus_dirs: &[PathBuf], objective_dir: PathBuf, broker_port: u16) -> Re
     let mut fuzzer = StdFuzzer::new(scheduler, feedback, objective);
 
     // The wrapped harness function, calling out to the LLVM-style harness
-    let mut harness = |buf: &[u8]| {
+    let mut harness = |input: &BytesInput| {
+        let buf = input.bytes();
```
Comment (domenukk) [inline fuzzers/libfuzzer_libmozjpeg/src/lib.rs:133]: `target_bytes`?

File: fuzzers/libfuzzer_libmozjpeg/src/lib.rs [hunk 3]
```diff
@@ -128,7 +129,8 @@ fn fuzz(corpus_dirs: &[PathBuf], objective_dir: PathBuf, broker_port: u16) -> Re
     let mut fuzzer = StdFuzzer::new(scheduler, feedback, objective);
 
     // The wrapped harness function, calling out to the LLVM-style harness
-    let mut harness = |buf: &[u8]| {
+    let mut harness = |input: &BytesInput| {
```
Comment (domenukk) [inline fuzzers/libfuzzer_libmozjpeg/src/lib.rs:132]: That could be `I: HasTargetBytes`, that way we support more Inputs automatically

File: fuzzers/libfuzzer_libmozjpeg/src/lib.rs (hunk 2 above)
Comment (domenukk) [reply fuzzers/libfuzzer_libmozjpeg/src/lib.rs:133]: @andreafioraldi does the split beween `bytes` and `target_bytes` makes sense atm? One was for the target, one for serialization?

File: fuzzers/baby_fuzzer/src/main.rs (hunk 1 above)
Comment (maxammann) [reply fuzzers/baby_fuzzer/src/main.rs:30]: Hm, Ill try and see what the rust compiler says. Maybe it needs a Sized type here. not sure.

File: fuzzers/libfuzzer_libmozjpeg/src/lib.rs (hunk 2 above)
Comment (maxammann) [reply fuzzers/libfuzzer_libmozjpeg/src/lib.rs:133]: Maybe we can remove `HasTargetBytes` and only use `HasBytesVec`, not sure though

File: fuzzers/baby_fuzzer/src/main.rs (hunk 1 above)
Comment (maxammann) [reply fuzzers/baby_fuzzer/src/main.rs:30]: Hm this would not work as the paramters needs to be an `Input` but input does not extend `HasBytesVec`

Comment (domenukk) [reply fuzzers/baby_fuzzer/src/main.rs:30]: In that case, why not `Input + HasTargetBytes`?

Comment (domenukk) [reply fuzzers/baby_fuzzer/src/main.rs:30]: Or, related, why does it need to be an `Input` in this case?

Comment (andreafioraldi) [reply fuzzers/baby_fuzzer/src/main.rs:30]: BytesInput is fine, the fuzzer use BytesInput

Comment (maxammann) [reply fuzzers/baby_fuzzer/src/main.rs:30]: I think in this case it is fine to use the concrete type BytesInput here as it is the concrete usage of the LibAFL library.

File: fuzzers/libfuzzer_libmozjpeg/src/lib.rs (hunk 2 above)
Comment (andreafioraldi) [reply fuzzers/libfuzzer_libmozjpeg/src/lib.rs]: No HasBytesVec is for Inputs that have a vec field and so returns a reference to these bytes. HasTargetBytes is for serialization to the bytes consumed by the target, and can be an owned type for the Inputs that produce the vector only when serializing (e.g. a grammar based Input)

File: fuzzers/baby_fuzzer/src/main.rs (hunk 1 above)
Comment (maxammann) [reply fuzzers/baby_fuzzer/src/main.rs:30]: Can't get it work:

```
error[E0225]: only auto traits can be used as additional traits in a trait object
  --> src/main.rs:30:45
   |
30 |     let mut harness = |input: &(dyn Input + HasTargetBytes)| {
   |                                     -----   ^^^^^^^^^^^^^^ additional non-auto trait
   |                                     |
   |                                     first non-auto trait
   |
   = help: consider creating a new trait with all of these as super-traits and using that trait here instead: `trait NewTrait: Input + HasTargetBytes {}`
   = note: auto-traits like `Send` and `Sync` are traits that have special properties; for more information on them, visit <https://doc.rust-lang.org/reference/special-types-and-traits.html#auto-traits>
```

**Comment by andreafioraldi** [comment]
Note for myself for later: this needs an update of the baby fuzzer page in the book

**Pull request merged**